x/auth/vesting/types/vesting_account.go

```diff
@@ -0,0 +1,619 @@
+package types
+
+import (
+	"errors"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	vestexported "github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	"gopkg.in/yaml.v2"
+)
+
+// Assert BaseVestingAccount implements authexported.Account at compile-time
+var _ authexported.Account = (*BaseVestingAccount)(nil)
+
+// Assert vesting accounts implement vestexported.VestingAccount at compile-time
+var _ vestexported.VestingAccount = (*ContinuousVestingAccount)(nil)
+var _ vestexported.VestingAccount = (*PeriodicVestingAccount)(nil)
+var _ vestexported.VestingAccount = (*DelayedVestingAccount)(nil)
+
+// Register the vesting account types on the auth module codec
+func init() {
+	authtypes.RegisterAccountTypeCodec(&BaseVestingAccount{}, "cosmos-sdk/BaseVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&ContinuousVestingAccount{}, "cosmos-sdk/ContinuousVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&DelayedVestingAccount{}, "cosmos-sdk/DelayedVestingAccount")
+	authtypes.RegisterAccountTypeCodec(&PeriodicVestingAccount{}, "cosmos-sdk/PeriodicVestingAccount")
+}
+
+// BaseVestingAccount implements the VestingAccount interface. It contains all
+// the necessary fields needed for any vesting account implementation.
+type BaseVestingAccount struct {
+	*authtypes.BaseAccount
+	OriginalVesting  sdk.Coins `json:"original_vesting" yaml:"original_vesting"`   // coins in account upon initialization
+	DelegatedFree    sdk.Coins `json:"delegated_free" yaml:"delegated_free"`       // coins that are vested and delegated
+	DelegatedVesting sdk.Coins `json:"delegated_vesting" yaml:"delegated_vesting"` // coins that vesting and delegated
+	EndTime          int64     `json:"end_time" yaml:"end_time"`                   // when the coins become unlocked
+}
+
+// NewBaseVestingAccount creates a new BaseVestingAccount object
+func NewBaseVestingAccount(baseAccount *authtypes.BaseAccount, originalVesting sdk.Coins,
+	endTime int64) *BaseVestingAccount {
+	return &BaseVestingAccount{
+		BaseAccount:      baseAccount,
+		OriginalVesting:  originalVesting,
+		DelegatedFree:    sdk.NewCoins(),
+		DelegatedVesting: sdk.NewCoins(),
+		EndTime:          endTime,
+	}
+}
+
+// SpendableCoinsFromVestingCoins returns all the spendable coins for a vesting account given a
+// set of vesting coins.
+//
+// CONTRACT: The account's coins, delegated vesting coins, vestingCoins must be
+// sorted.
+func (bva BaseVestingAccount) SpendableCoinsFromVestingCoins(vestingCoins sdk.Coins) sdk.Coins {
```

Comment:
Why the API name change?

Reply:
I think name change should be reverted. The current name is misleading in my opinion as it implies that this is calculating only the original vesting coins that have already vested. This isn't true.

If I send `50atoms` to a vesting account, it will be reflected in this function since `bc` will get incremented by `50atoms`

Reply:
The previous name, `spendableCoins`, can't simply be changed to `SpendableCoins` becuase it doesn't match the `SpendableCoins` signature defined on the `VestingAccount` interface. What would be a better name?